beacon-chain/p2p/peers/status.go

```diff
@@ -0,0 +1,311 @@
+// uncontactable due to being down, firewalled, etc.  Instead, this works with peers that are contactable but may or may not be of
+// the correct fork version, not currently required due to the number of current connections, etc.
+//
+// A peer can have one of a number of states:
+//
+// - disconnected if we are not able to talk to the remote peer
+// - connecting if we are attempting to be able to talk to the remote peer
+// - connected if we are able to talk to the remote peer
+// - disconnecting if we are attempting to stop being able to talk to the remote peer
+//
+// For convenience, there are two aggregate states expressed in functions:
+//
+// - active if we are connecting or connected
+// - inactive if we are disconnecting or disconnected
+//
+// Peer information is persistent for the run of the service.  This allows for collection of useful long-term statistics such as
+// number of bad responses obtained from the peer, giving the basis for decisions to not talk to known-bad peers.
+package peers
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/roughtime"
+)
+
+// PeerConnectionState is the state of the connection.
+type PeerConnectionState int
+
+const (
+	// PeerDisconnected means there is no connection to the peer.
+	PeerDisconnected PeerConnectionState = iota
+	// PeerConnecting means there is an on-going attempt to connect to the peer.
+	PeerConnecting
+	// PeerConnected means the peer has an active connection.
+	PeerConnected
+	// PeerDisconnecting means there is an on-going attempt to disconnect from the peer.
+	PeerDisconnecting
+)
+
+// Status is the structure holding the peer status information.
+type Status struct {
+	lock            sync.RWMutex
+	maxBadResponses int
+	status          map[peer.ID]*peerStatus
+}
+
+// peerStatus is the status of an individual peer at the protocol level.
+type peerStatus struct {
+	address               ma.Multiaddr
+	direction             network.Direction
+	peerState             PeerConnectionState
+	chainState            *pb.Status
+	chainStateLastUpdated time.Time
+	badResponses          int
+}
+
+// NewStatus creates a new status entity.
+func NewStatus(maxBadResponses int) *Status {
+	return &Status{
+		maxBadResponses: maxBadResponses,
+		status:          make(map[peer.ID]*peerStatus),
+	}
+}
+
+// MaxBadResponses returns the maximum number of bad responses a peer can provide before it is considered bad.
+func (p *Status) MaxBadResponses() int {
+	return p.maxBadResponses
+}
+
+// Add adds a peer.  This will error if the peer already exists.
+func (p *Status) Add(pid peer.ID, address ma.Multiaddr, direction network.Direction) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if _, ok := p.status[pid]; ok {
+		return fmt.Errorf("peer %v already exists", pid)
```

Comment:
I've made a couple of changes to the code.  With the latest commit, anything that sets data won't return an error as it will auto-create the underlying structures if required.  Anything that attempts to fetch data where the peer is not known will return a `ErrPeerUnknown` as per your suggestion.